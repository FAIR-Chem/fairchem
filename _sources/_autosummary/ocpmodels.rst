ocpmodels
=========

.. automodule:: ocpmodels

   
   
   

   
   
   

   
   
   

   
   
   



.. rubric:: Modules

.. autosummary::
   :toctree:
   :recursive:

   ocpmodels.common
   ocpmodels.datasets
   ocpmodels.models
   ocpmodels.modules
   ocpmodels.preprocessing
   ocpmodels.tasks
   ocpmodels.trainers

