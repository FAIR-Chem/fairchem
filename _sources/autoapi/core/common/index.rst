:py:mod:`core.common`
=====================

.. py:module:: core.common

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   relaxation/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   data_parallel/index.rst
   distutils/index.rst
   flags/index.rst
   gp_utils/index.rst
   hpo_utils/index.rst
   logger/index.rst
   registry/index.rst
   transforms/index.rst
   tutorial_utils/index.rst
   typing/index.rst
   utils/index.rst


