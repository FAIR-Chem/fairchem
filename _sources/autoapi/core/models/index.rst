:py:mod:`core.models`
=====================

.. py:module:: core.models


Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   equiformer_v2/index.rst
   escn/index.rst
   gemnet/index.rst
   gemnet_gp/index.rst
   gemnet_oc/index.rst
   painn/index.rst
   scn/index.rst
   utils/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   base/index.rst
   dimenet_plus_plus/index.rst
   model_registry/index.rst
   schnet/index.rst


Package Contents
----------------


Functions
~~~~~~~~~

.. autoapisummary::

   core.models.model_name_to_local_file



Attributes
~~~~~~~~~~

.. autoapisummary::

   core.models.available_pretrained_models


.. py:data:: available_pretrained_models

   

.. py:function:: model_name_to_local_file(model_name: str, local_cache: str | pathlib.Path) -> str

   Download a pretrained checkpoint if it does not exist already

   :param model_name: the model name. See available_pretrained_checkpoints.
   :type model_name: str
   :param local_cache: path to local cache directory
   :type local_cache: str or Path

   :returns: local path to checkpoint file
   :rtype: str


