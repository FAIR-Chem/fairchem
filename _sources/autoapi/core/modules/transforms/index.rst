:py:mod:`core.modules.transforms`
=================================

.. py:module:: core.modules.transforms


Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   core.modules.transforms.DataTransforms



Functions
~~~~~~~~~

.. autoapisummary::

   core.modules.transforms.decompose_tensor



.. py:class:: DataTransforms(config)


   .. py:method:: __call__(data_object)



.. py:function:: decompose_tensor(data_object, config) -> torch_geometric.data.Data


