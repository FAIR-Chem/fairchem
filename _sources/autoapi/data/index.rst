data
====

.. py:module:: data


Subpackages
-----------

.. toctree::
   :maxdepth: 1

   /autoapi/data/oc/index
   /autoapi/data/odac/index
   /autoapi/data/om/index


