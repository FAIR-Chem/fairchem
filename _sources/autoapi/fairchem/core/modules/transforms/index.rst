:py:mod:`fairchem.core.modules.transforms`
==========================================

.. py:module:: fairchem.core.modules.transforms


Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   fairchem.core.modules.transforms.DataTransforms



Functions
~~~~~~~~~

.. autoapisummary::

   fairchem.core.modules.transforms.decompose_tensor



.. py:class:: DataTransforms(config)


   .. py:method:: __call__(data_object)



.. py:function:: decompose_tensor(data_object, config) -> torch_geometric.data.Data


