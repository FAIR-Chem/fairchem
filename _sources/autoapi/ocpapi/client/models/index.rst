:py:mod:`ocpapi.client.models`
==============================

.. py:module:: ocpapi.client.models


Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   ocpapi.client.models._DataModel
   ocpapi.client.models.Model
   ocpapi.client.models.Models
   ocpapi.client.models.Bulk
   ocpapi.client.models.Bulks
   ocpapi.client.models.Adsorbates
   ocpapi.client.models.Atoms
   ocpapi.client.models.SlabMetadata
   ocpapi.client.models.Slab
   ocpapi.client.models.Slabs
   ocpapi.client.models.AdsorbateSlabConfigs
   ocpapi.client.models.AdsorbateSlabRelaxationsSystem
   ocpapi.client.models.AdsorbateSlabRelaxationsRequest
   ocpapi.client.models.Status
   ocpapi.client.models.AdsorbateSlabRelaxationResult
   ocpapi.client.models.AdsorbateSlabRelaxationsResults




.. py:class:: _DataModel


   Base class for all data models.

   .. py:attribute:: other_fields
      :type: dataclasses_json.CatchAll

      Fields that may have been added to the API that all not yet supported
      explicitly in this class.


.. py:class:: Model


   Bases: :py:obj:`_DataModel`

   Stores information about a single model supported in the API.

   .. py:attribute:: id
      :type: str

      The ID of the model.


.. py:class:: Models


   Bases: :py:obj:`_DataModel`

   Stores the response from a request for models supported in the API.

   .. py:attribute:: models
      :type: List[Model]

      The list of models that are supported.


.. py:class:: Bulk


   Bases: :py:obj:`_DataModel`

   Stores information about a single bulk material.

   .. py:attribute:: src_id
      :type: str

      The ID of the material.

   .. py:attribute:: formula
      :type: str

      The chemical formula of the material.

   .. py:attribute:: elements
      :type: List[str]

      The list of elements in the material.


.. py:class:: Bulks


   Bases: :py:obj:`_DataModel`

   Stores the response from a request to fetch bulks supported in the API.

   .. py:attribute:: bulks_supported
      :type: List[Bulk]

      List of bulks that can be used in the API.


.. py:class:: Adsorbates


   Bases: :py:obj:`_DataModel`

   Stores the response from a request to fetch adsorbates supported in the
   API.

   .. py:attribute:: adsorbates_supported
      :type: List[str]

      List of adsorbates that can be used in the API.


.. py:class:: Atoms


   Bases: :py:obj:`_DataModel`

   Subset of the fields from an ASE Atoms object that are used within this
   API.

   .. py:attribute:: cell
      :type: Tuple[Tuple[float, float, float], Tuple[float, float, float], Tuple[float, float, float]]

      3x3 matrix with unit cell vectors.

   .. py:attribute:: pbc
      :type: Tuple[bool, bool, bool]

      Whether the structure is periodic along the a, b, and c lattice vectors,
      respectively.

   .. py:attribute:: numbers
      :type: List[int]

      The atomic number of each atom in the unit cell.

   .. py:attribute:: positions
      :type: List[Tuple[float, float, float]]

      The coordinates of each atom in the unit cell, relative to the cartesian
      frame.

   .. py:attribute:: tags
      :type: List[int]

      Labels for each atom in the unit cell where 0 represents a subsurface atom
      (fixed during optimization), 1 represents a surface atom, and 2 represents
      an adsorbate atom.

   .. py:method:: to_ase_atoms() -> ase.Atoms

      Creates an ase.Atoms object with the positions, element numbers,
      etc. populated from values on this object.

      :returns: ase.Atoms object with values from this object.



.. py:class:: SlabMetadata


   Bases: :py:obj:`_DataModel`

   Stores metadata about a slab that is returned from the API.

   .. py:attribute:: bulk_src_id
      :type: str

      The ID of the bulk material from which the slab was derived.

   .. py:attribute:: millers
      :type: Tuple[int, int, int]

      The Miller indices of the slab relative to bulk structure.

   .. py:attribute:: shift
      :type: float

      The position along the vector defined by the Miller indices at which a
      cut was taken to generate the slab surface.

   .. py:attribute:: top
      :type: bool

      If False, the top and bottom surfaces for this millers/shift pair are
      distinct and this slab represents the bottom surface.


.. py:class:: Slab


   Bases: :py:obj:`_DataModel`

   Stores all information about a slab that is returned from the API.

   .. py:attribute:: atoms
      :type: Atoms

      The structure of the slab.

   .. py:attribute:: metadata
      :type: SlabMetadata

      Extra information about the slab.


.. py:class:: Slabs


   Bases: :py:obj:`_DataModel`

   Stores the response from a request to fetch slabs for a bulk structure.

   .. py:attribute:: slabs
      :type: List[Slab]

      The list of slabs that were generated from the input bulk structure.


.. py:class:: AdsorbateSlabConfigs


   Bases: :py:obj:`_DataModel`

   Stores the response from a request to fetch placements of a single
   absorbate on a slab.

   .. py:attribute:: adsorbate_configs
      :type: List[Atoms]

      List of structures, each representing one possible adsorbate placement.

   .. py:attribute:: slab
      :type: Slab

      The structure of the slab on which the adsorbate is placed.


.. py:class:: AdsorbateSlabRelaxationsSystem


   Bases: :py:obj:`_DataModel`

   Stores the response from a request to submit a new batch of adsorbate
   slab relaxations.

   .. py:attribute:: system_id
      :type: str

      Unique ID for this set of relaxations which can be used to fetch results
      later.

   .. py:attribute:: config_ids
      :type: List[int]

      The list of IDs assigned to each of the input adsorbate placements, in the
      same order in which they were submitted.


.. py:class:: AdsorbateSlabRelaxationsRequest


   Bases: :py:obj:`_DataModel`

   Stores the request to submit a new batch of adsorbate slab relaxations.

   .. py:attribute:: adsorbate
      :type: str

      Description of the adsorbate.

   .. py:attribute:: adsorbate_configs
      :type: List[Atoms]

      List of adsorbate placements being relaxed.

   .. py:attribute:: bulk
      :type: Bulk

      Information about the original bulk structure used to create the slab.

   .. py:attribute:: slab
      :type: Slab

      The structure of the slab on which adsorbates are placed.

   .. py:attribute:: model
      :type: str

      The type of the ML model being used during relaxations.

   .. py:attribute:: ephemeral
      :type: Optional[bool]

      Whether the relaxations can be deleted (assume they cannot be deleted if
      None).

   .. py:attribute:: adsorbate_reaction
      :type: Optional[str]

      If possible, an html-formatted string describing the reaction will be added
      to this field.


.. py:class:: Status(*args, **kwds)


   Bases: :py:obj:`enum.Enum`

   Relaxation status of a single adsorbate placement on a slab.

   .. py:attribute:: NOT_AVAILABLE
      :value: 'not_available'

      The configuration exists but the result is not yet available. It is
      possible that checking again in the future could yield a result.

   .. py:attribute:: FAILED_RELAXATION
      :value: 'failed_relaxation'

      The relaxation failed for this configuration.

   .. py:attribute:: SUCCESS
      :value: 'success'

      The relaxation was successful and the requested information about the
      configuration was returned.

   .. py:attribute:: DOES_NOT_EXIST
      :value: 'does_not_exist'

      The requested configuration does not exist.

   .. py:method:: __str__() -> str

      Return str(self).



.. py:class:: AdsorbateSlabRelaxationResult


   Bases: :py:obj:`_DataModel`

   Stores information about a single adsorbate slab configuration, including
   outputs for the model used in relaxations.

   The API to fetch relaxation results supports requesting a subset of fields
   in order to limit the size of response payloads. Optional attributes will
   be defined only if they are including the response.

   .. py:attribute:: config_id
      :type: int

      ID of the configuration within the system.

   .. py:attribute:: status
      :type: Status

      The status of the request for information about this configuration.

   .. py:attribute:: system_id
      :type: Optional[str]

      The ID of the system in which the configuration was originally submitted.

   .. py:attribute:: cell
      :type: Optional[Tuple[Tuple[float, float, float], Tuple[float, float, float], Tuple[float, float, float]]]

      3x3 matrix with unit cell vectors.

   .. py:attribute:: pbc
      :type: Optional[Tuple[bool, bool, bool]]

      Whether the structure is periodic along the a, b, and c lattice vectors,
      respectively.

   .. py:attribute:: numbers
      :type: Optional[List[int]]

      The atomic number of each atom in the unit cell.

   .. py:attribute:: positions
      :type: Optional[List[Tuple[float, float, float]]]

      The coordinates of each atom in the unit cell, relative to the cartesian
      frame.

   .. py:attribute:: tags
      :type: Optional[List[int]]

      Labels for each atom in the unit cell where 0 represents a subsurface atom
      (fixed during optimization), 1 represents a surface atom, and 2 represents
      an adsorbate atom.

   .. py:attribute:: energy
      :type: Optional[float]

      The energy of the configuration.

   .. py:attribute:: energy_trajectory
      :type: Optional[List[float]]

      The energy of the configuration at each point along the relaxation
      trajectory.

   .. py:attribute:: forces
      :type: Optional[List[Tuple[float, float, float]]]

      The forces on each atom in the relaxed structure.

   .. py:method:: to_ase_atoms() -> ase.Atoms

      Creates an ase.Atoms object with the positions, element numbers,
      etc. populated from values on this object.

      The predicted energy and forces will also be copied to the new
      ase.Atoms object as a SinglePointCalculator (a calculator that
      stores the results of an already-run simulation).

      :returns: ase.Atoms object with values from this object.



.. py:class:: AdsorbateSlabRelaxationsResults


   Bases: :py:obj:`_DataModel`

   Stores the response from a request for results of adsorbate slab
   relaxations.

   .. py:attribute:: configs
      :type: List[AdsorbateSlabRelaxationResult]

      List of configurations in the system, each representing one placement of
      an adsorbate on a slab surface.

   .. py:attribute:: omitted_config_ids
      :type: List[int]

      List of IDs of configurations that were requested but omitted by the
      server. Results for these IDs can be requested again.


