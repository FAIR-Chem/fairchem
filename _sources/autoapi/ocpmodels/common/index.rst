:py:mod:`ocpmodels.common`
==========================

.. py:module:: ocpmodels.common

.. autoapi-nested-parse::

   Copyright (c) Facebook, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   relaxation/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   data_parallel/index.rst
   distutils/index.rst
   flags/index.rst
   gp_utils/index.rst
   hpo_utils/index.rst
   logger/index.rst
   model_registry/index.rst
   registry/index.rst
   transforms/index.rst
   tutorial_utils/index.rst
   typing/index.rst
   utils/index.rst


