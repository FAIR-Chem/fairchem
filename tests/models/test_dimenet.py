import os

import numpy as np
import pytest
import torch
from ase.io import read
from torch_geometric.data import Batch, Data

from ocpmodels.common.transforms import RandomRotate
from ocpmodels.models import DimeNet
from ocpmodels.preprocessing import AtomsToGraphs


@pytest.fixture(scope="class")
def load_data(request):
    atoms = read(
        os.path.join(os.path.dirname(os.path.abspath(__file__)), "atoms.json"),
        index=0,
        format="json",
    )
    a2g = AtomsToGraphs(
        max_neigh=12,
        radius=6,
        dummy_distance=7,
        dummy_index=-1,
        r_energy=True,
        r_forces=True,
        r_distances=True,
    )
    data_list = a2g.convert_all([atoms])
    request.cls.data = data_list[0]


@pytest.fixture(scope="class")
def load_model(request):
    model = DimeNet(
        None, 32, 1, cutoff=6.0, regress_forces=True, use_pbc=False
    )
    request.cls.model = model


@pytest.mark.usefixtures("load_data")
@pytest.mark.usefixtures("load_model")
class TestDimeNet:
    def test_rotation_invariance(self):
        # Recreate the Data object to only keep the necessary features.
        data = Data(atomic_numbers=self.data.atomic_numbers, pos=self.data.pos)

        # Sampling a random rotation within [-180, 180] for all axes.
        transform = RandomRotate([-180, 180], [0, 1, 2])
        data_rotated, rot, inv_rot = transform(data.clone())
        assert not np.array_equal(data.pos, data_rotated.pos)

        # Pass it through the model.
        batch = Batch.from_data_list([data, data_rotated])
        out = self.model(batch)

        # Compare predicted energies and forces (after inv-rotation).
        energies = out[0].detach()
        np.testing.assert_almost_equal(energies[0], energies[1], decimal=5)

        forces = out[1].detach()
        np.testing.assert_array_almost_equal(
            forces[: forces.shape[0] // 2],
            torch.matmul(forces[forces.shape[0] // 2 :], inv_rot),
            decimal=5,
        )

    def test_energy_force_shape(self):
        # Recreate the Data object to only keep the necessary features.
        data = Data(atomic_numbers=self.data.atomic_numbers, pos=self.data.pos)

        # Pass it through the model.
        out = self.model(Batch.from_data_list([data]))

        # Compare shape of predicted energies, forces.
        energy = out[0].detach()
        np.testing.assert_equal(energy.shape, (1, 1))

        forces = out[1].detach()
        np.testing.assert_equal(forces.shape, (data.pos.shape[0], 3))
