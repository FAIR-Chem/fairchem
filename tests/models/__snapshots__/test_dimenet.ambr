# name: TestDimeNet.test_energy_force_shape
  Size(
    1,
    1,
  )
# ---
# name: TestDimeNet.test_energy_force_shape.1
  Approx(
  	array([[0.]], dtype=float32), 
  	rtol=1e-05, 
  	atol=1e-06
  )
# ---
# name: TestDimeNet.test_energy_force_shape.2
  Size(
    34,
    3,
  )
# ---
# name: TestDimeNet.test_energy_force_shape.3
  Approx(
  	array([[-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.],
  	       [-0., -0., -0.]], dtype=float32), 
  	rtol=1e-05, 
  	atol=1e-06
  )
# ---
